Symbolic execution fix

vladciobanu opened the pull request:
###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance: `stylish-haskell`

---

virgil-serbanuta commented on src/main/haskell/kore/test/Test/Kore/IndexedModule/MockMetadataTools.hs:
```diff
@@ -39,6 +40,7 @@ makeMetadataTools symbolOrAliasSorts attr isSubsortOf =
         , sortAttributes = const functionAttributes
         , symbolOrAliasSorts = symbolOrAliasSorts
         , isSubsortOf = \first second -> (first, second) `elem` isSubsortOf
+        , subsorts = const Set.empty
```
makeMetadataTools shoud get a table/graph/somehting on top of which it should compute `isSubsortOf` and `subsorts`, otherwise our code would get contradictory information in tests. Alternately, `isSubsortOf` could work as it currently does, but `makeMetadataTools` should also get a lookup table or something for `subsorts` and we would rely on the caller to keep them consistent.

virgil-serbanuta commented on src/main/haskell/kore/src/Kore/IndexedModule/MetadataTools.hs:
```diff
@@ -58,21 +66,39 @@ extractMetadataTools m =
+    , subsorts = Set.fromList . fmap getSortFromId . getSubsorts
     }
   where
+    subsortTable :: Map (Sort Object) [Sort Object]
```
You should add unit-tests for this.

vladciobanu replied:
Done; see https://github.com/kframework/kore/pull/298/commits/2ffd4b051b9fa1c141e1ba7d52e9555f049fcfbc#diff-c1602aa6b3c4f86ae4b16a7770174399R193 (and below)

virgil-serbanuta commented on src/main/haskell/kore/src/Kore/Step/Simplification/AndTerms.hs:
```diff
@@ -732,12 +733,19 @@ sortInjectionAndEqualsAssumesDifferentHeads
             then return (return (ExpandedPattern.bottom, SimplificationProof))
-            else
-                empty
+            else if Set.null sortIntersection
```
You should add unit-tests for this.

vladciobanu commented:
@virgil-serbanuta I did a few changes. First off, I realised that `isSubsortOf` is reflexive while `subsorts` did not include the requested sort in the resulting set, so I changed it to do so. This does not change the algorithm because the set intersection used in `AndTerms` still applies identically.

I added tests for both areas you mentioned -- see the replies to your review for exact links. Please let me know if you would rather i rebase before you re-check this. I think rebasing mid-review makes it harder to do subsequent reviews so I will not do it unless you specifically request it. The downside is, Jenkins will not run.
